sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -186,11 +186,11 @@ def task():
       try:
         self._execute(lambda: worker.do_instruction(work), work)
       finally:
-        # Delete the instruction_id <-> worker mapping
-        self._instruction_id_vs_worker.pop(work.instruction_id, None)
-        # Put the worker back in the free worker pool
-        self.workers.put(worker)
-
+        if not worker.has_active_processors():
```

Comment:
The bundle processors must be kept out of the free pool, but the workers need not be. Instead, grab a fresh worker whenever a finalization request comes through.

Reply:
If looking at here: https://github.com/apache/beam/blob/05bc3f6b8d46dbbdd6a17b19db01b92fc9c9e276/sdks/python/apache_beam/runners/worker/sdk_worker.py#L212, the SdkHarness is still caching the ```instruction_id -> worker``` map. With BundleProcessorCache, I think this map can be removed and when there is a new request coming in, get a new worker directly. wdyt?

Reply:
Implementation in commit https://github.com/apache/beam/pull/7937/commits/46dab24f559a62653f47a794336003e6530e9fc5

Reply:
Yes, this is great. 

The only issue is that there is a limited number of workers, and we don't want to use them up (or wait for them) for progress/split calls. Given that these worker are stateless and threadsafe, perhaps a reasonable solution would be to create one extra worker that we always use for progress and split (but not for process or bundle finalization). 

Other than that, looks good.

Reply:
Setup a thread_pool with one thread for finalization call only: https://github.com/apache/beam/pull/7937/commits/c748ecfc43d11856720e843a115200323bc96586